lib/srv/server/cloudwatcher.go

```diff
@@ -0,0 +1,195 @@
+	ticker := time.NewTicker(w.fetchInterval)
+	defer ticker.Stop()
+	for {
+		for _, fetcher := range w.fetchers {
+			inst, err := fetcher.GetEC2Instances(w.ctx)
+			if err != nil {
+				if trace.IsNotFound(err) {
+					continue
+				}
+				log.WithError(err).Error("Failed to fetch EC2 instances")
+				continue
+			}
+			select {
+			case w.InstancesC <- *inst:
+			case <-w.ctx.Done():
+			}
+		}
+		select {
+		case <-ticker.C:
+			continue
+		case <-w.ctx.Done():
+			return
+		}
+	}
+}
+
+// Stop stops the watcher
+func (w *Watcher) Stop() {
+	w.cancel()
+}
+
+// NewCloudWatcher creates a new cloud watcher instance.
+func NewCloudWatcher(ctx context.Context, matchers []services.AWSMatcher, clients cloud.Clients) (*Watcher, error) {
+	cancelCtx, cancelFn := context.WithCancel(ctx)
+	watcher := Watcher{
+		fetchers:      []*ec2InstanceFetcher{},
+		ctx:           cancelCtx,
+		cancel:        cancelFn,
+		fetchInterval: time.Minute,
+		InstancesC:    make(chan EC2Instances),
+	}
+	for _, matcher := range matchers {
+		for _, region := range matcher.Regions {
+			cl, err := clients.GetAWSEC2Client(region)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			fetcher := newEC2InstanceFetcher(ec2FetcherConfig{
+				Matcher:   matcher,
+				Region:    region,
+				Document:  matcher.SSM.DocumentName,
+				EC2Client: cl,
+				Labels:    matcher.Tags,
+			})
+			watcher.fetchers = append(watcher.fetchers, fetcher)
+		}
+	}
+	return &watcher, nil
+}
+
+type ec2FetcherConfig struct {
+	Matcher   services.AWSMatcher
+	Region    string
+	Document  string
+	EC2Client ec2iface.EC2API
+	Labels    types.Labels
+}
+
+type ec2InstanceFetcher struct {
+	Filters    []*ec2.Filter
+	EC2        ec2iface.EC2API
+	Region     string
+	Document   string
+	Parameters map[string]string
+}
+
+func newEC2InstanceFetcher(cfg ec2FetcherConfig) *ec2InstanceFetcher {
+	tagFilters := []*ec2.Filter{&ec2.Filter{
+		Name:   aws.String(constants.AWSInstanceStateName),
+		Values: aws.StringSlice([]string{ec2.InstanceStateNameRunning}),
+	}}
+
+	for key, val := range cfg.Labels {
+		tagFilters = append(tagFilters, &ec2.Filter{
+			Name:   aws.String("tag:" + key),
+			Values: aws.StringSlice(val),
+		})
+	}
+	fetcherConfig := ec2InstanceFetcher{
+		EC2:      cfg.EC2Client,
+		Filters:  tagFilters,
+		Region:   cfg.Region,
+		Document: cfg.Document,
+		Parameters: map[string]string{
+			"token": cfg.Matcher.Params.JoinToken,
+		},
+	}
+	return &fetcherConfig
+}
+
+// GetEC2Instances fetches all EC2 instances matching configured filters.
+func (f *ec2InstanceFetcher) GetEC2Instances(ctx context.Context) (*EC2Instances, error) {
+	var instances []*ec2.Instance
+	var accountID string
+	err := f.EC2.DescribeInstancesPagesWithContext(ctx, &ec2.DescribeInstancesInput{
+		Filters: f.Filters,
+	},
+		func(dio *ec2.DescribeInstancesOutput, b bool) bool {
+			for _, res := range dio.Reservations {
+				if accountID == "" {
+					accountID = aws.StringValue(res.OwnerId)
```

Comment:
What is the use of `AccountID`?  If my understanding is correct, the `OwnerId` is the owner of the reservation.